### PR TITLE
Priority queue

### PR DESCRIPTION
**Pull request by danking**

### CODE REVIEW

File: src/main/scala/is/hail/utils/BinaryHeap.scala [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    val temp = a(i)
+    a(i) = a(j)
+    a(j) = temp
+    m(a(j).v) = j
+    m(a(i).v) = i
+  }
+
+  private def maybeGrow() {
+    if (next >= a.length) {
+      val a2 = new Array[RankedT](a.length << 1)
```
Comment (jigold) [inline src/main/scala/is/hail/utils/BinaryHeap.scala]: See comment above for arrayBuilder. If decide to keep this method, use `Array.copy` instead.

File: src/main/scala/is/hail/utils/BinaryHeap.scala [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  private def maybeShrink() {
+    if (next < (a.length >>> 2)) {
+      val a2 = new Array[RankedT](a.length >>> 2)
+      var j = 0
+      while (j < a2.length) {
+        a2(j) = a(j)
+        j += 1
+      }
```
Comment (jigold) [inline src/main/scala/is/hail/utils/BinaryHeap.scala]: Use Array.copy.

File: src/main/scala/is/hail/utils/BinaryHeap.scala [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      while (j < a.length) {
+        a2(j) = a(j)
+        j += 1
+      }
+      a = a2
+    }
+  }
+
+  private def maybeShrink() {
+    if (next < (a.length >>> 2)) {
```
Comment (jigold) [inline src/main/scala/is/hail/utils/BinaryHeap.scala]: Should this be `>>> 1` analogous to `maybeGrow`?

File: src/main/scala/is/hail/utils/BinaryHeap.scala [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  private val m: mutable.Map[T, Int] = new mutable.HashMap()
+  private var next: Int = 0
+
+  def size: Int = next
+
+  def insert(t: T, r: Long) {
+    if (m.contains(t))
+      throw new RuntimeException(s"key $t already exists with priority ${m(t)}, cannot add it again with priority $r")
+    maybeGrow()
+    put(next, new RankedT(t,r))
```
Comment (jigold) [inline src/main/scala/is/hail/utils/BinaryHeap.scala]: space between parameters `(t, r)`.

File: src/main/scala/is/hail/utils/BinaryHeap.scala [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    a(i).rank = r
+    bubbleUp(i)
+  }
+
+  def contains(t: T): Boolean =
+    m.contains(t)
+
+  def toArray: Array[T] = {
+    val trimmed = new Array(size)
+    var i = 0
```
Comment (jigold) [inline src/main/scala/is/hail/utils/BinaryHeap.scala]: Use `Array.copy`.

File: src/test/scala/is/hail/utils/BinaryHeapSuite.scala [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    assert(bh.contains(1) == true)
+    bh.insert(2, 10)
+    assert(bh.contains(2) == true)
+    assert(bh.max() === 2)
+    assert(bh.size === 2)
+    assert(bh.max() === 2)
+    assert(bh.size === 2)
+    assert(bh.extractMax() === 2)
+    assert(bh.contains(2) == false)
+    assert(bh.size === 1)
```
Comment (jigold) [inline src/test/scala/is/hail/utils/BinaryHeapSuite.scala]: Why are you repeating the checks of the `size`? I assume you are repeating `max` to ensure it's not popping the max element.

File: src/main/scala/is/hail/utils/BinaryHeap.scala [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+class BinaryHeap[@specialized T : ClassTag](initialCapacity: Int = 32) {
+  private class RankedT(val v: T, var rank: Long) {}
+  private var a: Array[RankedT] = new Array[RankedT](initialCapacity)
+  private val m: mutable.Map[T, Int] = new mutable.HashMap()
+  private var next: Int = 0
+
+  def size: Int = next
```
Comment (jigold) [inline src/main/scala/is/hail/utils/BinaryHeap.scala]: Can you add an `isEmpty`-like method?

File: src/test/scala/is/hail/utils/BinaryHeapSuite.scala [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+    bh.insert(1, 0)
+    bh.insert(2, 0)
+    bh.insert(3, 0)
+    bh.insert(4, 0)
+    bh.insert(5, 0)
+    assert(true)
+  }
+
+  @Test
+  def growPastCapacity32() {
```
Comment (jigold) [inline src/test/scala/is/hail/utils/BinaryHeapSuite.scala]: Do you test the shrink capacity functionality?

**Comment by jigold** [comment]
Also, can you add a larger test with 1024 elements? Maybe check 557th element is correct after 556 extractMax operations?

File: src/main/scala/is/hail/utils/BinaryHeap.scala (hunk 3 above)
Comment (danking) [inline src/main/scala/is/hail/utils/BinaryHeap.scala]: if it was perfectly symmetric, then if the queue keeps changing sizes from 31 to 32 then we'd be constantly increasing and decreasing the size of the array. With this strategy, I have to do an order of magnitude of operations to trigger another size change. (i.e. if I double the buffer to 64 by putting something at position 32, I now have to do 16 units of work before I shrink the buffer).

File: src/main/scala/is/hail/utils/BinaryHeap.scala (hunk 7 above)
Comment (danking) [inline src/main/scala/is/hail/utils/BinaryHeap.scala]: yes

File: src/main/scala/is/hail/utils/BinaryHeap.scala (hunk 5 above)
Comment (danking) [inline src/main/scala/is/hail/utils/BinaryHeap.scala]: done

File: src/main/scala/is/hail/utils/BinaryHeap.scala (hunk 1 above)
Comment (danking) [inline src/main/scala/is/hail/utils/BinaryHeap.scala]: done

File: src/main/scala/is/hail/utils/BinaryHeap.scala (hunk 2 above)
Comment (danking) [inline src/main/scala/is/hail/utils/BinaryHeap.scala]: done

File: src/test/scala/is/hail/utils/BinaryHeapSuite.scala (hunk 6 above)
Comment (danking) [inline src/test/scala/is/hail/utils/BinaryHeapSuite.scala]: Yeah I guess I repeat max to make sure I get the same value and I repeat the size checks to make sure the size hasn't changed. I can see that this is a bit redundant. I can remove if you prefer.

Comment (jigold) [inline src/test/scala/is/hail/utils/BinaryHeapSuite.scala]: Maybe check it once?

File: src/main/scala/is/hail/utils/BinaryHeap.scala [hunk 9]
```diff
@@ -0,0 +1,154 @@
+package is.hail.utils
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+
+class BinaryHeap[@specialized T : ClassTag](initialCapacity: Int = 32) {
+  private class RankedT(val v: T, var rank: Long) {}
+  private var a: Array[RankedT] = new Array[RankedT](initialCapacity)
```
Comment (danking) [inline src/main/scala/is/hail/utils/BinaryHeap.scala]: I thought about this, and I'd want to add random access writing to ArrayBuilder, @jigold @cseed  thoughts on this change? I also like my shrinking strategy 😉, but I could go with ArrayBuilder if shrinking is not supported by practical experience.

File: src/test/scala/is/hail/utils/BinaryHeapSuite.scala (hunk 6 above)
Comment (danking) [inline src/test/scala/is/hail/utils/BinaryHeapSuite.scala]: I like it. Done.

File: src/test/scala/is/hail/utils/BinaryHeapSuite.scala (hunk 8 above)
Comment (danking) [inline src/test/scala/is/hail/utils/BinaryHeapSuite.scala]: I do now :)

**Comment by danking** [comment]
@cseed I believe I addressed jackie's comments, can you take a look?

**Comment by danking** [comment]
done, done, and done.

**Comment by danking** [comment]
@jigold back to you

File: src/main/scala/is/hail/utils/BinaryHeap.scala [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+  }
+
+  def checkHeapProperty() {
+    checkHeapProperty(0)
+  }
+
+  private def checkHeapProperty(current: Int) {
+    val leftChild = (current << 1) + 1
+    val rightChild = (current << 1) + 2
+    if (leftChild < next) {
```
Comment (jigold) [inline src/main/scala/is/hail/utils/BinaryHeap.scala]: Sorry, I should have been more clear. Is there a reason you can't just do this:

```
if (leftChild < next) {
assertHeapProperty(leftChild, current)
checkHeapProperty(leftChild)
}

if (rightChild < next) {
assertHeapProperty(rightChild, current)
checkHeapProperty(rightChild)
}
```

Comment (danking) [reply src/main/scala/is/hail/utils/BinaryHeap.scala]: I have a vague sense that the highest heap property violation is the most interesting one, so I want to `assertHeapProperty` on the `rightChild` before checking the heap property down the left side of the tree.

**Pull request merged**